File: dotnet/Base/OpenStack.PolyIO/System.IO/Polyfill.cs
using System.Numerics;
using System.Runtime.InteropServices;

namespace System.IO;

public static partial class Polyfill {
    #region Lump

    // lumps
    [StructLayout(LayoutKind.Sequential)]
    public struct X_LumpON { public int Offset; public int Num; }

    [StructLayout(LayoutKind.Sequential)]
    public struct X_LumpNO { public int Num; public int Offset; }

    [StructLayout(LayoutKind.Sequential)]
    public struct X_LumpNO2 { public int Num; public int Offset; public int Offset2; }

    [StructLayout(LayoutKind.Sequential)]
    public struct X_Lump2NO { public int Num; public int Offset; public int Offset2; }

    #endregion

    #region BoundBox

    [StructLayout(LayoutKind.Sequential)]
    public struct X_BoundBox {
        public Vector3 Min;                // minimum values of X,Y,Z
        public Vector3 Max;                // maximum values of X,Y,Z
    }

    #endregion
}

File: rust/core/polyio/src/system_io/polyfill.rs
// PORT-SOURCE: Core/OpenStack.PolyIO/System.IO/Polyfill.cs
// PORT-SHA: 348ffb11dac35500
// PORT-STATUS: done
//
// The on-disk header structs shared across formats. C# marks these
// `[StructLayout(LayoutKind.Sequential)]` so they can be blitted; `#[repr(C)]`
// is the exact equivalent. They stay POD and `Copy`, as in C#.
//
// Note these are *little-endian on disk* in every current caller. Reading them
// via `BinaryReaderExt` field-by-field (rather than transmuting) keeps that
// explicit and portable to big-endian hosts.

/// C# `X_LumpON` — offset then count.
#[repr(C)]
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct XLumpON {
    pub offset: i32,
    pub num: i32,
}

/// C# `X_LumpNO` — count then offset.
#[repr(C)]
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct XLumpNO {
    pub num: i32,
    pub offset: i32,
}

/// C# `X_LumpNO2` — count then two offsets.
#[repr(C)]
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct XLumpNO2 {
    pub num: i32,
    pub offset: i32,
    pub offset2: i32,
}

/// C# `X_Lump2NO`.
///
/// Field-for-field identical to [`XLumpNO2`] in the C# source; both are kept so
/// the two trees stay diffable, but if that turns out to be a copy-paste slip
/// upstream, collapse them in both trees at once.
#[repr(C)]
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct XLump2NO {
    pub num: i32,
    pub offset: i32,
    pub offset2: i32,
}

/// C# `X_BoundBox` — axis-aligned min/max.
#[repr(C)]
#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct XBoundBox {
    /// Minimum values of X, Y, Z.
    pub min: [f32; 3],
    /// Maximum values of X, Y, Z.
    pub max: [f32; 3],
}
